crux-lucene/src/crux/lucene.clj

```diff
@@ -170,83 +176,128 @@
+  [{:keys [^Path db-dir bus tx-log document-store index-store analyzer indexer query-engine] :as opts}]
   (let [directory (FSDirectory/open db-dir)
         index-writer (->index-writer directory analyzer)
         searcher-manager (SearcherManager. index-writer false false nil)
-        lucene-store (LuceneNode. directory analyzer index-writer searcher-manager indexer)]
-    ;; Ensure lucene index exists for immediate queries:
-    (.commit index-writer)
-    (validate-lucene-store-up-to-date index-store lucene-store)
+        !ingester-error (atom nil)
+        lucene-store (map->LuceneNode {:directory directory
+                                       :analyzer analyzer
+                                       :index-writer index-writer
+                                       :searcher-manager searcher-manager
+                                       :indexer indexer
+                                       :!ingester-error !ingester-error})]
+
+    (.commit index-writer) ; ensure lucene index exists for immediate queries
+
     (q/assoc-pred-ctx! query-engine ::lucene-store lucene-store)
-    (bus/listen bus {:crux/event-types #{:crux.tx/committing-tx :crux.tx/aborting-tx}
-                     :crux.bus/executor (reify java.util.concurrent.Executor
-                                          (execute [_ f]
-                                            (.run f)))}
-                (fn [ev]
-                  (when (= :crux.tx/committing-tx (:crux/event-type ev))
-                    (index! indexer index-writer (db/fetch-docs document-store (:doc-ids ev)))
-                    (when-let [evicting-eids (not-empty (:evicting-eids ev))]
-                      (evict! indexer index-writer evicting-eids)))
-                  (index-tx! index-writer (:submitted-tx ev))
-                  (.commit index-writer)
-                  (.maybeRefreshBlocking searcher-manager)))
-    lucene-store))
+
+    (-> lucene-store
+        (assoc :ingester-fut
+               (db/subscribe-async tx-log (db/latest-completed-tx lucene-store)
+                                   (let [await-deps (select-keys opts #{:bus :tx-ingester})]
+                                     (fn [_acc {:keys [::txe/tx-events ::tx/tx-id] :as tx}]
+                                       (try
+                                         (tx/await-tx await-deps tx)
+
+                                         (let [committed? (not (db/tx-failed? index-store tx-id))]
+                                           (when committed?
+                                             (let [{:keys [docs evicted-eids]} (transform-tx-events document-store tx-events)]
+                                               (index-tx! indexer index-writer evicted-eids docs)))
+
+                                           (complete-tx! index-writer tx)
+                                           (.commit index-writer) ; TODO aware of #1500
+                                           (.maybeRefreshBlocking searcher-manager)
```

Comment:
On reflection, it is now perhaps best to do this work in the new `await-tx` - WDYT?

Reply:
it's an option - although someone (you?) mentioned in the call earlier not to put all this work on one unfortunate reader?

(assuming you mean `.maybeRefreshBlocking`? given I'm pretty sure `complete-tx!` has to stay here (and it's quick enough anyway), and `.commit`'s not required for readers, can do that on a schedule)

Reply:
(Yep, just meant the `.maybeRefreshBlocking`, sorry :))

> you, I think?

I did yep, quoting from https://lucene.apache.org/core/7_0_0/core/org/apache/lucene/search/SearcherManager.html
> In addition you should periodically call ReferenceManager.maybeRefresh(). While it's possible to call this just before running each query, this is discouraged since it penalizes the unlucky queries that need to refresh. It's better to use a separate background thread, that periodically calls ReferenceManager.maybeRefresh().

The benefit of putting it on await-tx is that it will make the ingest/catch-up even snappier (which might well be _very_ significant, it's hard to judge at this stage) and also then the user has complete control to schedule their own periodic (not immediately before a query) `l/await-tx` _if_ they want to avoid penalizing random queries.

Reply:
fair. will leave it as-is for this PR, #1500 seems better placed to handle it :smile: